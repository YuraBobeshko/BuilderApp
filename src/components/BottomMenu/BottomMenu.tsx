import React, { useState, useEffect } from "react";
import { ITypes } from "../../types";
const options = [{ name: "component" }, { name: "folder" }, { name: "file" }];

export interface IBottomMenu {
  name?: string;
  text?: string | Promise<string>;
  type?: ITypes;
  closeOrOpen?: "open" | "close";
  setName: (name: string) => void;
  setText?: (name: string) => void;
  onClickAdd: () => void;
  onClickDelete?: () => void;
  onChangeType: (type: ITypes) => void;
  onClickCloseOrOpen: () => void;
}

const BottomMenu = ({
  name,
  text,
  type,
  closeOrOpen,
  setName,
  setText,
  onChangeType,
  onClickAdd,
  onClickDelete,
  onClickCloseOrOpen,
}: IBottomMenu) => {
  const [openTextEditor, setOpenTextEditor] = useState(false);
  const [textRes, setTextRes] = useState("");
  // @ts-ignore
  useEffect(async () => {
    if (setText)
      if (text instanceof Promise) {
        const data = await text;
        setTextRes(data || '');
      } else if (text) {
        setTextRes(text);
      }
  }, [openTextEditor, text, setText]);

  return (
    <>
      <div>
        <button className='add' onClick={onClickAdd}>add</button>
        <button className='delete' onClick={onClickDelete}>delete</button>
        <button className='closeOrOpen' onClick={onClickCloseOrOpen} disabled={!closeOrOpen}>
          {closeOrOpen || "not parent"}
        </button>
        <select
          className='selectType'
          onChange={(
            event: React.ChangeEvent<HTMLSelectElement> & {
              target: { value: ITypes };
            }
          ) => onChangeType(event.target.value)}
          value={type || "select"}
        >
          <option value="select" hidden>
            select
          </option>
          {options.map(({ name }) => (
            <option key={name} value={name}>
              {name}
            </option>
          ))}
        </select>
        {setText && (
          <button className='setText' onClick={() => setOpenTextEditor(!openTextEditor)}>
            {openTextEditor ? "closeTextEditor" : "openTextEditor"}
          </button>
        )}
        <input className='name' value={name} onChange={(e) => setName(e.target.value)} />
      </div>
      {openTextEditor && setText && (
        <textarea
          className='text'
          onChange={(e) => setText(e.target.value)}
          value={textRes}
          style={{ width: 500, height: 200 }}
        />
      )}
      {textRes}
    </>
  );
};

export default BottomMenu;
