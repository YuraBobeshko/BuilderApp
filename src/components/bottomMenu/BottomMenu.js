import React, { useState } from "react";

const options = [{ name: "component" }, { name: "folder" }, { name: "file" }];

const BottomMenu = ({
  name,
  text,
  type,
  closeOrOpen,
  setName,
  onChangeType,
  onClickAdd,
  onClickDelet,
  onClickCloseOrOpen,
}) => {
  const [openTextEditor, setOpenTextEditor] = useState(false);
  return (
    <>
      <div>
        <button onClick={onClickAdd}>add</button>
        <button onClick={onClickDelet}>delete</button>
        <button onClick={onClickCloseOrOpen} disabled={!!onClickCloseOrOpen}>
          {closeOrOpen || "not parent"}
        </button>
        <select
          onChange={(e) => {
            onChangeType(e.target.value);
          }}
          value={type || "select"}
        >
          <option value="select" hidden>
            select
          </option>
          {options.map(({ name }) => (
            <option key={name} value={name}>
              {name}
            </option>
          ))}
        </select>
        <button onClick={() => setOpenTextEditor(!openTextEditor)}>
          {openTextEditor ? "closeTextEditor" : "openTextEditor"}
        </button>
        <input value={name} onChange={(e) => setName(e.target.value)} />
      </div>
      {openTextEditor && (
        <textarea
          onChange={() => {}}
          value={text}
          style={{ width: 500, height: 200 }}
        />
      )}
    </>
  );
};

export default BottomMenu;
