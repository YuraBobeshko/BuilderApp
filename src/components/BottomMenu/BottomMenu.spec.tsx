import 'jsdom-global/register';
import BottomMenu, {IBottomMenu} from "./BottomMenu";
import {mount, shallow} from "enzyme";
import {act, waitFor} from "@testing-library/react";

const defaultProps: IBottomMenu = {
    onChangeType: () => {
    },
    onClickAdd: () => {
    },
    onClickCloseOrOpen: () => {
    },
    setName: () => {
    }
}

const value = 'component'

function args() {
    return {
        target: {
            value
        }
    }
}

const setUp = (props?: Partial<IBottomMenu>) => shallow(
    <BottomMenu
        {...defaultProps}
        {...props}
    />
);

const setUpMount = (props?: Partial<IBottomMenu>) => mount(
    <BottomMenu
        {...defaultProps}
        {...props}
    />
);


describe("BottomMenu component", () => {
    let component = setUp();
    beforeEach(() => {
        component = setUp();
    })

    it("should call onClick method add btn", () => {
        const mockCallBack = jest.fn();
        const component = setUp({
            onClickAdd: mockCallBack
        });
        expect(mockCallBack.mock.calls.length).toBe(0)
        component.find('.add').simulate('click');
        expect(mockCallBack.mock.calls.length).toBe(1)
    });

    it("should call onClick method delete btn", () => {
        const mockCallBAck = jest.fn();
        const component = setUp({
            onClickDelete: mockCallBAck
        });
        expect(mockCallBAck.mock.calls.length).toBe(0)
        component.find('.delete').simulate('click');
        expect(mockCallBAck.mock.calls.length).toBe(1)
    });

    it("should call onClick method closeOrOpen btn", () => {
        const mockCallBack = jest.fn();
        const component = setUp({
            onClickCloseOrOpen: mockCallBack
        });
        expect(mockCallBack.mock.calls.length).toBe(0)
        component.find('.closeOrOpen').simulate('click');
        expect(mockCallBack.mock.calls.length).toBe(1)
    });

    it("should call onChange method nameInput", () => {
        const mockCallBack = jest.fn()
        const component = setUp({setName: mockCallBack})
        const input = component.find('.name');
        input.simulate('change', args())
        expect(mockCallBack.mock.calls[0][0]).toBe(value)
    });

    it("should set text as string", () => {
        const mockCallBack = jest.fn()
        const component = setUpMount({setText: mockCallBack})
        component.find('.setText').simulate('click')
        const textarea = () => component.find('.text');

        textarea().simulate('change', args())
         component.setProps({text: mockCallBack.mock.calls[0][0]})

        component.update()
        expect(textarea().props().value).toBe(value)
    });

    it("should set text as promise", async () => {
        const mockCallBack = jest.fn()
        const component = setUpMount({setText: mockCallBack})
        component.find('.setText').simulate('click')
        await act(async() => {
            await  component.setProps({text: Promise.resolve(value)});
        })
        await act(async () => {
            await component.update();
        })

        return expect(component.find('.text').props().value).toBe(value)
    });

    it("should call onChange method selectType", () => {
        const mockCallBack = jest.fn()
        const component = setUp({onChangeType: mockCallBack})
        const input = component.find('.selectType');
        input.simulate('change', args())
        expect(mockCallBack.mock.calls[0][0]).toBe(value)
    });

    describe('logics work with setText', () => {
        let mockCallBack = jest.fn()
        let component = setUp({setText: mockCallBack})

        beforeEach(() => {
            mockCallBack = jest.fn()
            component = setUp({setText: mockCallBack})
        })

        it("should render button setText", () => {
            expect(component.find('.setText').length).toBe(1)
        });

        it("should show textarea after click on btn.setTest", () => {
            expect(mockCallBack.mock.calls.length).toBe(0)
            component.find('.setText').simulate('click')
            expect(component).toMatchSnapshot()
        });
    })

    it('should shallow with min count of props', () => {
        expect(component).toMatchSnapshot()
    })

    it('should shallow with max count of props', () => {
        const component = setUp({
            name: 'g',
            text: 'f',
            type: 'file',
            closeOrOpen: 'open'
        });

        expect(component).toMatchSnapshot()
    })
});
