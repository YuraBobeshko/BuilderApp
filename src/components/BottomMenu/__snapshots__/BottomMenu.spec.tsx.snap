// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomMenu component logics work with setText should show textarea after click on btn.setTest 1`] = `
<Fragment>
  <div>
    <button
      className="add"
      onClick={[Function]}
    >
      add
    </button>
    <button
      className="delete"
    >
      delete
    </button>
    <button
      className="closeOrOpen"
      disabled={true}
      onClick={[Function]}
    >
      not parent
    </button>
    <select
      className="selectType"
      onChange={[Function]}
      value="select"
    >
      <option
        hidden={true}
        value="select"
      >
        select
      </option>
      <option
        key="component"
        value="component"
      >
        component
      </option>
      <option
        key="folder"
        value="folder"
      >
        folder
      </option>
      <option
        key="file"
        value="file"
      >
        file
      </option>
    </select>
    <button
      className="setText"
      onClick={[Function]}
    >
      closeTextEditor
    </button>
    <input
      className="name"
      onChange={[Function]}
    />
  </div>
  <textarea
    className="text"
    onChange={[Function]}
    style={
      Object {
        "height": 200,
        "width": 500,
      }
    }
    value=""
  />
</Fragment>
`;

exports[`BottomMenu component should shallow with max count of props 1`] = `
<Fragment>
  <div>
    <button
      className="add"
      onClick={[Function]}
    >
      add
    </button>
    <button
      className="delete"
    >
      delete
    </button>
    <button
      className="closeOrOpen"
      disabled={false}
      onClick={[Function]}
    >
      open
    </button>
    <select
      className="selectType"
      onChange={[Function]}
      value="file"
    >
      <option
        hidden={true}
        value="select"
      >
        select
      </option>
      <option
        key="component"
        value="component"
      >
        component
      </option>
      <option
        key="folder"
        value="folder"
      >
        folder
      </option>
      <option
        key="file"
        value="file"
      >
        file
      </option>
    </select>
    <input
      className="name"
      onChange={[Function]}
      value="g"
    />
  </div>
</Fragment>
`;

exports[`BottomMenu component should shallow with min count of props 1`] = `
<Fragment>
  <div>
    <button
      className="add"
      onClick={[Function]}
    >
      add
    </button>
    <button
      className="delete"
    >
      delete
    </button>
    <button
      className="closeOrOpen"
      disabled={true}
      onClick={[Function]}
    >
      not parent
    </button>
    <select
      className="selectType"
      onChange={[Function]}
      value="select"
    >
      <option
        hidden={true}
        value="select"
      >
        select
      </option>
      <option
        key="component"
        value="component"
      >
        component
      </option>
      <option
        key="folder"
        value="folder"
      >
        folder
      </option>
      <option
        key="file"
        value="file"
      >
        file
      </option>
    </select>
    <input
      className="name"
      onChange={[Function]}
    />
  </div>
</Fragment>
`;
